File: boj/21937.cpp
#include <bits/stdc++.h>
using namespace std;
#define FASTIO cin.tie(nullptr)->sync_with_stdio(false)

int N, M, X;
vector<int> a[100001];
bool visited[100001];
int ans = 0;

void dfs(int cur, int step) {
    visited[cur] = true;
    for (int prev : a[cur]) {
        if (!visited[prev]) {
            ++ans;
            dfs(prev, step + 1);
        }
    }
}

int main() {
    FASTIO;
    cin >> N >> M;
    for (int i = 0; i < M; ++i) {
        int u, v;
        cin >> u >> v;
        a[v].push_back(u);
    }
    cin >> X;
    dfs(X, 0);
    cout << ans;
    return 0;
}

File: boj/21937.py
import sys
sys.setrecursionlimit(10**6)
si = sys.stdin.readline

N, M = map(int, si().split())
a = [[] for _ in range(N + 1)]
visited = [False] * (N + 1)
ans = 0

for _ in range(M):
    u, v = map(int, si().split())
    a[v].append(u)

X = int(si())


def dfs(cur):
    global ans
    visited[cur] = True
    for prev in a[cur]:
        if not visited[prev]:
            ans += 1
            dfs(prev)


dfs(X)
print(ans)
